Add intelX scan to Pe_source

DJensen94 opened the pull request:
Fixes #352 
Integrate new data source (intelX) into data gathering scans
## 🗣 Description ##
This PR adds a new data source scan to pull credential breaches identified by IntelligenceX
## 💭 Motivation and context ##
This scan is important to add because our current credential breach scans have scaling limitations. This source has no limitations other than api response times, we can use it on any domain for any organization and get results. I will also find additional breaches that our current scans do not find.


## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [X] This PR has an informative and human-readable title.
- [X] Changes are limited to a single goal - *eschew scope creep!*
- [X] *All* future TODOs are captured in issues, which are referenced
      in code comments.
- [ ] All new and existing tests pass.

DJensen94 committed:
Add intelX scan to Pe_source

Integrate new data source (intelX) into data gathering scans

DJensen94 committed:
update db_query_source import

Update import call to match correct file

DJensen94 committed:
remove module that hasn't been added yet

remove DNSMonitor which hasn't been added to develop yet

DJensen94 committed:
Add get_params function

Add function to pull api parameters from relevant files

DJensen94 committed:
remove global variable

remove unnecessary global variable

DJensen94 committed:
add intelx to database.ini

add section for intelx to the database.ini

DJensen94 committed:
fix section name

fix intelx capitalization in section name

DJensen94 committed:
remove soc_med_included flag

remove flag for filtering values from data gathering

DJensen94 committed:
add else statement to source logic

If a valid source is not provided program should exit scan

dav3r commented on src/pe_reports/data/database.ini:
```diff
@@ -29,3 +29,6 @@ client_secret=
 [whoisxml]
+
+[intelx]
+api_key=
```
This section should be moved up to maintain alphabetical order of the sections.

DJensen94 replied:
@dav3r Issue addressed in the following commit: [3284942](https://github.com/cisagov/pe-reports/pull/414/commits/328494231a43ca6a39327a59c90b14bd4bb77c63)

dav3r commented on src/pe_source/data/pe_db/db_query_source.py:
```diff
@@ -48,7 +48,7 @@ def get_orgs():
     try:
         cur = conn.cursor()
-        sql = """SELECT * FROM organizations"""
+        sql = """SELECT * FROM organizations where report_on or demo"""
```
You may want to update the doc-string for this function.  "Query organizations table" is not particularly specific, and now there's a new `where` clause that you might want to mention.

DJensen94 replied:
@dav3r Issue addressed in the following commit: [d2851ea](https://github.com/cisagov/pe-reports/pull/414/commits/d2851eacac1c1cad3fe811a2d52b7e2985aa1f12)

dav3r replied:
👍 Much better, though I noted a small misspelling [here](https://github.com/cisagov/pe-reports/pull/414/files#r1149423864).

dav3r commented on src/pe_source/data/pe_db/db_query_source.py:
```diff
@@ -417,3 +416,82 @@ def org_root_domains(conn, org_uid):
+
+
+def insert_intelx_breaches(df):
+    """Insert intelx breach data."""
```
Use consistent capitalization:
```suggestion
    """Insert IntelX breach data."""
```

dav3r commented on src/pe_source/data/pe_db/db_query_source.py:
```diff
@@ -417,3 +416,82 @@ def org_root_domains(conn, org_uid):
+
+
+def insert_intelx_credentials(df):
+    """Insert sixgill credential data."""
```
Copy/paste error?
```suggestion
    """Insert IntelX credential data."""
```

dav3r commented on src/pe_source/data/pe_db/db_query_source.py:
```diff
@@ -417,3 +416,82 @@ def org_root_domains(conn, org_uid):
+            tuples,
+        )
+        conn.commit()
+        LOGGER.info("Successfully inserted/updated breaches into PE database.")
```
Consider making this log message more specific:
```suggestion
        LOGGER.info("Successfully inserted/updated IntelX breaches into PE database.")
```

dav3r commented on src/pe_source/data/pe_db/db_query_source.py:
```diff
@@ -417,3 +416,82 @@ def org_root_domains(conn, org_uid):
+        )
+        conn.commit()
+        LOGGER.info(
+            "Successfully inserted/updated exposed credentials into PE database."
```
Consider making this log message more specific:
```suggestion
            "Successfully inserted/updated exposed IntelX credentials into PE database."
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+"""Collect Intelx credential leak data."""
```
Use consistent capitalization:
```suggestion
"""Collect IntelX credential leak data."""
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+    org_root_domains,
+)
+
+# Calculate Datetimes for collection period
```
Unnecessary capitalization:
```suggestion
# Calculate datetimes for collection period
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+        self.orgs_list = orgs_list
+
+    def run_intelx(self):
+        """Run IntelX api calls."""
```
Use proper capitalization:
```suggestion
        """Run IntelX API calls."""
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+        try:
+            insert_intelx_breaches(breaches_df)
+        except Exception as e:
+            LOGGER.error("Failed inserting breaches for %s", cyhy_org_id)
```
Consider making this log message more specific:
```suggestion
            LOGGER.error("Failed inserting IntelX breaches for %s", cyhy_org_id)
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+        try:
+            insert_intelx_credentials(creds_df)
+        except Exception as e:
+            LOGGER.error("Failed inserting credentials for %s", cyhy_org_id)
```
Consider making this log message more specific:
```suggestion
            LOGGER.error("Failed inserting IntelX credentials for %s", cyhy_org_id)
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+                time.sleep(5)
+                attempts += 1
+                if attempts == 5:
+                    LOGGER.error("IntelX identity is not responding. Exiting program.")
```
Use consistent capitalization:
```suggestion
                    LOGGER.error("IntelX Identity is not responding. Exiting program.")
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+                    sys.exit()
+                LOGGER.info("IntelX Identity API response timed out. Trying again.")
+            except Exception as e:
+                LOGGER.error(f"Error occured geting search id: {e}")
```
Fix spelling errors:
```suggestion
                LOGGER.error(f"Error occurred getting search id: {e}")
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+                sys.exit()
+            LOGGER.info("IntelX Identity API response timed out. Trying again.")
+        except Exception as e:
+            LOGGER.error(f"Error occured geting search results: {e}")
```
Fix spelling errors:
```suggestion
            LOGGER.error(f"Error occurred getting search results: {e}")
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+            time.sleep(5)
+            attempts += 1
+            if attempts == 5:
+                LOGGER.error("IntelX identity is not responding. Exiting program.")
```
Use consistent capitalization:
```suggestion
                LOGGER.error("IntelX Identity is not responding. Exiting program.")
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+
+        payload = {}
+        headers = {}
+        attempts = 0
+        try:
+            response = requests.request("GET", url, headers=headers, data=payload)
+        except requests.exceptions.Timeout:
+            time.sleep(5)
+            attempts += 1
+            if attempts == 5:
```
I think you're missing a `while` loop here?

DJensen94 replied:
@dav3r fixed issue in the following commit: [257c7b8](https://github.com/cisagov/pe-reports/pull/414/commits/257c7b8be23b2620f662933038bbed5cef906570)

dav3r replied:
[`while attempts < 5:`](https://github.com/cisagov/pe-reports/blob/2b9e3eaaddca5609045a8330d61705c849488b08/src/pe_source/intelx_identity.py#L132) isn't really adding anything here- you could simply go with `while True:` since you already exit the loop `if attempts == 5`.

DJensen94 replied:
@dav3r issue fixed in the following commit: [f23b485](https://github.com/cisagov/pe-reports/pull/414/commits/f23b485c4d8d78aacb2c726f058ad80f07617d31)

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+                    if current_results:
+                        # Add the root_domain to each result object
+                        LOGGER.info(
+                            f"Intelx returned {len(current_results)} more credentials for {domain}"
```
Use consistent capitalization:
```suggestion
                            f"IntelX returned {len(current_results)} more credentials for {domain}"
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+                    time.sleep(3)
+                # If still waiting on new results wait
+                elif results["status"] == 1:
+                    LOGGER.info("Intelx still searching for more credentials")
```
Use consistent capitalization:
```suggestion
                    LOGGER.info("IntelX still searching for more credentials")
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+                    if current_results:
+                        # Add the root_domain to each result object
+                        LOGGER.info(
+                            f"Intelx returned {len(current_results)} more credentials for {domain}"
```
Use consistent capitalization:
```suggestion
                            f"IntelX returned {len(current_results)} more credentials for {domain}"
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,289 @@
+            roots_df["root_domain"].values.tolist(), START_DATE, END_DATE
+        )
+        if len(leaks_json) < 1:
+            LOGGER.info(f"No credentials found for {cyhy_org_id}")
```
We avoid using f-strings in logging statements - see [here](https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/) for some good reasons why:
```suggestion
            LOGGER.info("No credentials found for %s", cyhy_org_id)
```

I noticed a bunch other instances of using f-strings in logging messages in this PR - please modify them as well (I don't want to flag them all now, but I'll check for them before I give my approval).

DJensen94 replied:
@dav3r  other instances fixed in the following commit: [2b9e3ea](https://github.com/cisagov/pe-reports/pull/414/commits/2b9e3eaaddca5609045a8330d61705c849488b08)

dav3r reviewed (changes requested):
You'll notice a few trends in my feedback, though there are a couple of unique ones in there as well.

DJensen94 committed:
Alphabetize sections in database.ini

move intelx section up to maintain alphabetized order

DJensen94 committed:
update doc string in get_orgs function

update doc string to reflect the new where clause in the get_orgs function

DJensen94 committed:
Fix capitalization to be consistent

Fix capitalization of IntelX to match the rest of the instances.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix copy/paste error

Fix copy and paste error with wrong service name in doc string

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add clarification to logging statement

Clarify what scan is inserting data in the logging statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

dav3r commented on src/pe_source/data/pe_db/db_query_source.py:
```diff
@@ -44,11 +44,11 @@ def close(conn):
 
 def get_orgs():
-    """Query organizations table."""
+    """Query organizations that recieve reports and demo organizations."""
```
Fix misspelling:
```suggestion
    """Query organizations that receive reports and demo organizations."""
```

DJensen94 committed:
Add more description to log statement

specify the service in the logging statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix capitalization of IntelX in doc string

Keep consistent capitalization of IntelX

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix capitalization of "datetimes"

update capitalization where unnecessary

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix capitalization of API

Fix API capitalization when in all under case

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add service name in logging message

Add IntelX service name into its logging statement to make it clear where the script is running

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add service name in logging message

Add IntelX service name to logging message.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Update capitalization to be consistent

Update "IntelX Identity" to have consistent capitalization.

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix spelling error in logging statement

Fix spelling errors in logging statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix spelling error in logging statement

Fix spelling error in logging statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Maintain consistent capitalization.

Fix capitalization in logging statement to be consistent

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Add missing while loop

Add missing while loop to get_search_results function

DJensen94 committed:
Capitalize IntelX in logging statement

Fix capitalization of IntelX in log statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix capitalization of IntelX to be consistent

Fix capitalization of IntelX to be consistent

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix misspelling of the word receive

fix misspelling of "receive"

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix IntelX capitalization

Fix capitalization of IntelX in log statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove f string from log statement

Remove f string from log statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove f strings from logging statements

Remove all f strings from logging statements

dav3r commented:
@DJensen94 FYI, once you update this branch with the latest code in `develop`, this branch should be able to pass the [linting tests that are currently failing](https://github.com/cisagov/pe-reports/actions/runs/4534814363/jobs/7989471524?pr=414).

DJensen94 committed:
Change while criteria

Since the while loop has exit logic already the while clause just needs to be "while True:"

DJensen94 committed:
Merge branch 'develop' into DJ-add_intelx-dev

DJensen94 commented:
@dav3r i've made the suggested changes and merged develop into this branch

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,296 @@
+                    if current_results:
+                        # Add the root_domain to each result object
+                        LOGGER.info(
+                            "Intelx returned %s more credentials for %s",
```
Looks like we missed one:
```suggestion
                            "IntelX returned %s more credentials for %s",
```

dav3r commented on src/pe_source/intelx_identity.py:
```diff
@@ -0,0 +1,296 @@
+                    if current_results:
+                        # Add the root_domain to each result object
+                        LOGGER.info(
+                            "Intelx returned %s more credentials for %s",
```
Or maybe more than one 😄 
```suggestion
                            "IntelX returned %s more credentials for %s",
```

DJensen94 committed:
Fix capitalization of IntelX

Fix capitalization to keep it consistent

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix capitalization in log statement

Fix capitalization of service name in log statement

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

cduhn17 commented on src/pe_source/data/pe_db/config.py:
```diff
@@ -82,3 +82,21 @@ def cybersix_token():
+            params = parser.items(section)
+        else:
+            raise Exception(
+                "Section {} not found in the {} file".format(section, REPORT_DB_CONFIG)
```
Do we want to follow the same '%s' string replacement field methods like we do in the rest of the code?

dav3r replied:
That's up to you.  The main thing is that logging statements use printf %-style formatting and not f-strings - see [here](https://blog.pilosus.org/posts/2020/01/24/python-f-strings-in-logging/) for an explanation.  For exceptions and pretty much everywhere else, I generally prefer to use f-strings.

DJensen94 replied:
the linter reformats that back to how we currently have it, so as long as it isn't in a logging statement, I think we are ok as we currently have it.

cduhn17 reviewed (commented):
See comment at src/pe_source/data/pe_db/config.py.

dav3r reviewed (approved):
Thanks, as always, for considering my feedback! 🚀

cduhn17 reviewed (approved):
LGTM
